src/Channel.cpp

```diff
@@ -84,7 +87,92 @@ void Channel::remove(Channel *c) {
 	QWriteLocker lock(&c_qrwlChannels);
 	c_qhChannels.remove(c->iId);
 }
-#endif
+
+void Channel::setFilterMode(ChannelFilter filterMode) {
+	m_filterMode        = filterMode;
+	ServerHandlerPtr sh = Global::get().sh;
+	if (sh) {
+		Global::get().db->setChannelFilterMode(sh->qbaDigest, iId, m_filterMode);
+	}
+}
+
+void Channel::clearFilterMode() {
+	if (m_filterMode != ChannelFilter::NORMAL) {
+		setFilterMode(ChannelFilter::NORMAL);
+	}
+}
+
+bool Channel::isFiltered() const {
+	if (!Global::get().s.bFilterActive) {
+		// Channel filtering is disabled
+		return false;
+	}
+
+	const Channel *userChannel = nullptr;
+
+	if (Global::get().uiSession != 0) {
+		const ClientUser *user = ClientUser::get(Global::get().uiSession);
+		if (user) {
+			userChannel = user->cChannel;
+		}
+	}
+
+	bool hasUser = false;
+	bool hasHide = false;
+	bool hasPin  = false;
+
+	// Iterate tree down
+	std::queue< const Channel * > channelSearch;
+	channelSearch.push(this);
+
+	while (!channelSearch.empty()) {
+		const Channel *c = channelSearch.front();
+		channelSearch.pop();
+
+		// Never hide channel, if user resides in this channel or a subchannel
+		if (userChannel && c == userChannel) {
+			return false;
+		}
+
+		if (c->m_filterMode == ChannelFilter::PIN) {
+			hasPin = true;
+		}
+
+		if (!c->qlUsers.isEmpty()) {
+			hasUser = true;
+		}
+
+		for (const Channel *currentSubChannel : c->qlChannels) {
+			channelSearch.push(currentSubChannel);
+		}
+	}
+
+	// Iterate tree up
+	const Channel *c = this;
+	while (c) {
+		if (c->m_filterMode == ChannelFilter::HIDE) {
+			hasHide = true;
+			break;
+		}
+		c = c->cParent;
+	}
+
+	if (hasPin) {
+		return false;
+	}
+
+	if (hasHide) {
+		return true;
+	}
+
+	if (Global::get().s.bFilterHidesEmptyChannels) {
+		return !hasUser;
+	}
+
+	return false;
```

Comment:
```suggestion
	return Global::get().s.bFilterHidesEmptyChannel && !hasUser;
```
This is shorter and saves an additional branching (though that probably hardl matters in this particular code part), but perhaps harder to read. I'll leave the decision of whether or not to apply this patch up to you.